### PR TITLE
[WIP] Add Dirichlet-multinomial distribution.

### PR DESCRIPTION
**Pull request by bsmith89**
Dirichlet multinomial distribution.

The self.random implementation is non-standard, not well tested, and may be broken, and the documentation is currently lacking.  But the log-likelihood has been working for me for a few years without obvious issues.

### CODE REVIEW

**Comment by twiecki** [comment]
How can this be tested?

**Comment by junpenglao** [comment]
Need test +1. I think we can follow the test for BetaBinomial here.

**Comment by junpenglao** [comment]
See:
https://github.com/pymc-devs/pymc3/blob/e81df2d19ddc4066648b4b2dfc72431c6824f96f/pymc3/tests/test_distributions_random.py#L390-L392
https://github.com/pymc-devs/pymc3/blob/e81df2d19ddc4066648b4b2dfc72431c6824f96f/pymc3/tests/test_distributions_random.py#L604-L606
https://github.com/pymc-devs/pymc3/blob/e81df2d19ddc4066648b4b2dfc72431c6824f96f/pymc3/tests/test_distributions.py#L717-L718

**Comment by ColCarroll** [comment]
We should probably fix this, but you'll also have to add a line in `pymc3/distributions/__init__.py` to import this...

File: pymc3/distributions/multivariate.py [hunk 1]
```diff
@@ -722,6 +723,75 @@ def _repr_latex_(self, name=None, dist=None):
...
+
+    ==========  ===========================================
+    Support     :math:`x \in \{0, 1, \ldots, n\}` such that
+                :math:`\sum x_i = n`
+    Mean        :math:`n \frac{\alpha_i}{\sum{\alpha_k}}`
+    ==========  ===========================================
+
+    Parameters
+    ----------
+    alpha : one- or two-dimensional array
```
Comment (rpgoldman) [inline pymc3/distributions/multivariate.py]: As far as I can tell, sphinx won't format these correctly with the `:` set apart by spaces.  I think you need to change these to, for example `alpha: one- or two-dimensional array` or the online docs will come out wrong.

**Comment by rpgoldman** [comment]
> We should probably fix this, but you'll also have to add a line in `pymc3/distributions/__init__.py` to import this...

Why do we have the assignment to `__add__` in this file, if we aren't going to simply do `from .multinomial import *` in `distributions/__init__.py`?

File: pymc3/distributions/multivariate.py (hunk 1 above)
Comment (bsmith89) [reply pymc3/distributions/multivariate.py]: I'm matching the style found in all of the other docstrings for distribution classes.  Is this wrong?

**Comment by ColCarroll** [comment]
@rpgoldman -- that's exactly what we should do, but it is not in the scope of this PR, I think.

File: pymc3/distributions/multivariate.py (hunk 1 above)
Comment (ColCarroll) [reply pymc3/distributions/multivariate.py]: i agree all the other docs in this file are doing this. I think it is ok if you leave it, and we can file an issue to fix these all in one go.

Comment (ColCarroll) [reply pymc3/distributions/multivariate.py]: (@rpgoldman if that's ok with you)

**Comment by bsmith89** [comment]
Should I rebase this branch with master as I develop the tests, and, if so, how do I then update the PR?

**Comment by ColCarroll** [comment]
You don't have to, but it reduces the chance of merge conflicts. 

Yeah, if you rebase off master and push to your branch, it will update the PR, and run the test suite on CI. 

Please do ping when it is ready for another review!

**Comment by bsmith89** [comment]
Just pushed some (relatively messy) commits that include at least basic tests.

(@ColCarroll and others)

Right now I can see a few reasons this PR is still a WIP, but I'd welcome comments.

1.  This implementation of the DM distribution only handles a 1d vector for `n` and a 2d vector for `alpha`.  Some users may want to be able to specify a scalar `n` or higher numbers of dimensions.  I'm having trouble figuring out best to make it polymorphic over possible dimensions of these parameters.
2.  This implementation seems to require that the shape passed explicitly to `__init__`, even if it could be inferred from `n.shape` and `alpha.shape`.  It's not clear to me what best practices are for passing the inferred shape to `super().__init__`.
3.  Tests are not as extensive as for other distributions.

Nonetheless, this implementation seems to do what it says on the label.  I'd welcome feedback, but may not have time in the next month or two to implement the fancy, shape-handling logic that I see for e.g. the Multinomial distribution.

Happy to squash and clean up some commits if that's wanted.

**Comment by twiecki** [comment]
This looks pretty good to merge, what do you think @ColCarroll?

**Comment by AlexAndorra** [comment]
This would still be a nice addition IMO! What should we do to push it over the finish line?
Just rebase on master and let CI run, or are there still some important features to add?

File: pymc3/distributions/multivariate.py [hunk 2]
```diff
@@ -709,6 +710,85 @@ def logp(self, x):
...
+        result = const + series.sum(axis=-1)
+        return bound(result,
+                     tt.all(tt.ge(x, 0)),
+                     tt.all(tt.gt(alpha, 0)),
+                     tt.all(tt.ge(n, 0)),
+                     tt.all(tt.eq(x.sum(axis=-1), n)),
+                     broadcast_conditions=False)
+
+    def random(self, point=None, size=None, repeat=None):
+        alpha, n = draw_values([self.alpha, self.n], point=point)
```
Comment (ColCarroll) [inline pymc3/distributions/multivariate.py]:
```suggestion
        alpha, n = draw_values([self.alpha, self.n], point=point, size=size)
```

**Comment by twiecki** [comment]
Closing in favor of #4373.

**Pull request closed without merging**